### ISSUE TITLE
`with_bootsnap_fallback`: Constants are loaded twice when a `NameError` occurs

### ISSUE THREAD

**Issue by xuorig** [Shopify/bootsnap#74]
### Problem

`NameError` during `load_missing_constant` results in constants being loaded twice.

### Investigation

Take the following basic example, which can be found in [this repo](https://github.com/xuorig/bootsnap-issue/blob/master/example.rb) also:

```ruby
# example.rb
a = Bar.new
```

```ruby
# Bar.rb
class Bar
  DoesNotExist
end
```
 
  The following methods are called when `Bar` is first encountered:

  - `ActiveSupport::Dependencies#const_missing `
  - `Bootsnap#load_missing_constant`
  - `Bootsnap#with_bootsnap_fallback`
  - `super` => `ActiveSupport::load_missing_constant from_mod: Object, const_name: Bar`
  -  ...

Once Bar is loaded, `DoesNotExist` now has to be loaded, the same methods are called for `DoesNotExist` but remember this is all wrapped in the original `Bootsnap#load_missing_constant` and `Bootsnap#with_bootsnap_fallback`.

Since `DoesNotExist` cannot be found, a `NameError` is raised. It is reraised through the stack until it hits the rescue in [with_bootsnap_fallback](https://github.com/Shopify/bootsnap/blob/master/lib/bootsnap/load_path_cache/core_ext/active_support.rb#L7-L12).

The exception is caught, and then `yield` is called again because the loading is tried this time `without_bootsnap_cache`. This means the orignial `load_missing_constant` for `DoesNotExist` is called again. This of cource raises a `NameError` again, which this time is not caught by this rescue. Instead, it is caught by an other `load_missing_constant` call, the original call to load the `Bar` class.

The same process is repeated, `yield` is called again, which causes `Bar` to be reloaded, executing any logic twice.

### Real life example

This causes issues when a loaded constant sets state on load. Take this example:

```ruby
class Foo
  class_attribute :items
  self.items = {}
  
  def self.add_item(name, klass)
   if items.key?(name)
     raise Error, "cannot add same item twice!!!"
   end
   
    self.items = items.merge(name: klass)
  end
  
  # This one will run ok
  add_item :a, Bar
  # This one raises and causes the double load
  add_item :b, DoesNotExist
end
```

When loading `DoesNotExist`, `DoesNotExist` is being loaded twice, which fails both times but `Foo` is also loaded twice since it rescues the `NameError`.

This causes `add_item, :a, Bar` to be called twice, which causes this error to be raised:

`raise Error, "cannot add same item twice!!!"`

In reality, we would've liked to see the `NameError` raised, which causes a really weird developer experience.

### TL;DR

In `load_missing_constant`, `NameErrors` are caught by Bootsnap's `with_bootsnap_fallback`, and retried without bootsnap cache. This causes ActiveSupport's `load_missing_constant` to be called twice on all constants that were loading / led to the `NameError`.

### Questions

  - Is this the expected behaviour?
  - My ruby internals knowledge is limited on this one, should constants be loaded twice like this, could we unload them before trying without bootsnap cache?
  - Is this something developers need to be mindful of? Seems like this could happen to anybody using the gem.

cc @jules2689, @burke, @rafaelfranca

**Comment by xuorig** [comment]
btw i'd love to contribute a fix if I can get a bit more context from all of you 💃

**Comment by jules2689** [comment]
Good overview of what is happening:
![image](https://user-images.githubusercontent.com/3074765/28072094-e3c91e0c-661f-11e7-9311-e6fe3ccb2866.png)

**Comment by jules2689** [comment]
So the issue i see here is that the `without_bootsnap_cache` causes the class to be reloaded, which is normally fine. Since we have a DSL that modifies class attributes like this (with `add_item` and a class_attribute of `items`) we are experiencing double mutation of that entity.

It is important to note that this is only an issue when the class-level manipulation of objects fails with a NameError

**Comment by burke** [comment]
I don't have a solution for this offhand. It might not be super-complex, but I don't have time to dive in right now. Definitely seems like an unfortunate issue, but it also seems like unless you're doing something very wrong, it only ever makes a legitimate error less useful.
